tests: extend combined_infra_metrics test to cover ap metrics query

arnout opened the pull request:
Use the check_cmdu functionality to validate that the ap metrics responses are correct.

While we're at it, check the combined infra metrics as well.
Turns out they are wrong - an issue will be created for it.

Lots of supporting changes to the sniffer infrastructure.

arnout committed:
test_flows.py: factor policy_config_metric_reporting into a function

The send-and-check-response flow of setting the metric reporting policy
can be used in other tests as well. Therefore, factor it in a separate
function. Make the agent and STA reporting policy bits configurable.

While we're at it, change the check into a check_cmdu instead of
check_log.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: add check_cmdu_type_single wrapper

In many cases, we expect one and only one CMDU with certain properties.
Add a function check_cmdu_type_single for this. It returns the single
packet found, or None.

Update the existing check_cmdu_type calls to use this instead.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tests: sniffer.py: add comparison function to Tlv

For some tests, it is convenient to be able to check that a TLV is equal
to an expected TLV. Add a comparison operator to be able to do this. It
simply compares the relevant elements of the dictionary.

Factor the dictionary creation out into the function _d().

Note that it would probably better to store the complete dictionary in a
private member and add a __getattr__ member, but that's a bigger change.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
test_flows.py: add check_cmdu_has_tlv functions

Add two helper functions for checking the contents of a TLV:
- check_cmdu_has_tlvs checks if the CMDU has a TLV of a certain type.
- check_cmdu_has_tlv_single additionally checks that ther is only one.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

rmelotte commented on agent/src/beerocks/monitor/monitor_stats.cpp:
```diff
@@ -14,6 +14,7 @@
 
 #include <beerocks/tlvf/beerocks_message.h>
 #include <beerocks/tlvf/beerocks_message_monitor.h>
+#include <tlvf/wfa_map/tlvApMetrics.h>
```
It looks like this commit comes from https://github.com/prplfoundation/prplMesh/pull/1285
Is this PR going to be rebased on 1285 after 1285 is merged?

rmelotte commented on tests/test_flows.py:
```diff
@@ -133,6 +133,52 @@ def check_cmdu_type_single(
+            debug("  {}".format(packet))
+            return None
+        return tlvs[0]
+
```
To me these 2 functions clearly look like something the Packet class should be doing (`Packet.has_tlvs` and `Packet.has_tlv_single` or something like that).
Is there a specific reason to have them here?

arnout replied:
The idea is that the "check" style functions are part of the test_flows class (which should really be a base class that other classes derive from, but that's for later). If we strip down the has_tlvs function to something that could be part of the Packet class, it would just be `return [tlv for tlv in packet.ieee1905_tlvs if tlv.tlv_type == tlv_type]` - and we'd still need a check function in test_flows.py to do the actual check (i.e. to indicate failure if the result is empty).

rmelotte commented on tests/test_flows.py:
```diff
@@ -553,32 +553,99 @@ def test_link_metric_query(self):
+                debug("AP metrics 1: {}".format(ap_metrics_1))
+                debug("AP metrics 2: {}".format(ap_metrics_2))
+
+        def check_equal_tlv(tlv_type, expected, msg):
```
The combined naming of `check_equal_tlv` and `check_tlv_equal` is really confusing.
What about something like `check_has_tlv` for `check_equal_tlv`?

arnout replied:
I have instead added a function check_cmdu_has_tlvs_exact that does the job in a much more concise and readable way.

rmelotte reviewed (approved):
I didn't review the commit "agent: backhaul_manager: Implement AP Metrics Query/Response message" (4f77d2757bc53bcc6b28657862c1880539e63a73) , I will review it in the other PR (https://github.com/prplfoundation/prplMesh/pull/1285/commits)
Edit: scratch that, it looks like this PR has this one included:  https://github.com/prplfoundation/prplMesh/pull/1231/
So it just needs to be rebased on master AFAICS.

arnout committed:
test_flows.py: combined_infra_metrics: extend to cover AP metrics query

While we're at it, check the packets sent around instead of checking the
log files.

Some of the checks are commented out because they are not implemented
yet.

This also shows a bug in the combined infra metrics, so that check is
commented out as well.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
